ledger/cow.go

```diff
@@ -126,15 +128,43 @@ func (cb *roundCowState) getCreator(cidx basics.CreatableIndex, ctype basics.Cre
 	return cb.lookupParent.getCreator(cidx, ctype)
 }
 
-func (cb *roundCowState) lookup(addr basics.Address) (data basics.AccountData, err error) {
+func (cb *roundCowState) lookup(addr basics.Address) (pad ledgercore.PersistedAccountData, err error) {
 	d, ok := cb.mods.Accts.Get(addr)
 	if ok {
-		return d, nil
+		return ledgercore.PersistedAccountData{AccountData: d}, nil
 	}
 
 	return cb.lookupParent.lookup(addr)
 }
 
+// lookupWithHolding is gets account data but also fetches asset holding or app local data for a specified creatable
+func (cb *roundCowState) lookupHolding(addr basics.Address, cidx basics.CreatableIndex, ctype basics.CreatableType) (data ledgercore.PersistedAccountData, err error) {
+	ad, modified := cb.mods.Accts.Get(addr)
+	if modified {
+		exist := false
+		if ctype == basics.AssetCreatable {
+			_, exist = ad.Assets[basics.AssetIndex(cidx)]
+		} else {
+			_, exist = ad.AppLocalStates[basics.AppIndex(cidx)]
+		}
+
+		if exist {
+			return ledgercore.PersistedAccountData{AccountData: ad}, nil
+		}
+	}
+
+	d, err := cb.lookupParent.lookupHolding(addr, cidx, ctype)
```

Comment:
is this always needed ? i.e. in case the `ad` already contain the holding, there is no need to lookup, right ?

Reply:
yes, if ad in `mods` and the asset exists, it is returned above